clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -307,34 +377,346 @@ impl FunctionType {
         };
         check_argument_count(expected_args.len(), func_args)?;
 
-        for (expected_arg, arg) in expected_args.iter().zip(func_args.iter()).into_iter() {
-            match (&expected_arg.signature, arg) {
-                (
-                    TypeSignature::TraitReferenceType(trait_id),
-                    Value::Principal(PrincipalData::Contract(contract)),
-                ) => {
-                    let contract_to_check = db
-                        .load_contract(contract)
-                        .ok_or_else(|| CheckErrors::NoSuchContract(contract.name.to_string()))?;
-                    let trait_definition = db
-                        .get_defined_trait(&trait_id.contract_identifier, &trait_id.name)
-                        .unwrap()
-                        .ok_or(CheckErrors::NoSuchContract(
-                            trait_id.contract_identifier.to_string(),
-                        ))?;
-                    contract_to_check.check_trait_compliance(trait_id, &trait_definition)?;
-                }
-                (expected_type, value) => {
-                    if !expected_type.admits(&value) {
-                        let actual_type = TypeSignature::type_of(&value);
-                        return Err(
-                            CheckErrors::TypeError(expected_type.clone(), actual_type).into()
-                        );
+        let mut arg_types = Vec::new();
+        for arg in func_args {
+            arg_types.push(self.principal_to_callable_type(arg, 1)?);
+        }
+
+        for (expected_arg, arg_type) in expected_args.iter().zip(arg_types.iter()).into_iter() {
+            inner_type_check_type(
+                db,
+                None,
+                arg_type,
+                &expected_arg.signature,
+                clarity_version,
+                1,
+                &mut LimitedCostTracker::new_free(),
+            )?;
+        }
+        Ok(returns.clone())
+    }
+}
+
+pub fn check_functions_compatible<T: CostTracker>(
```

Comment:
Sorry to bug you on this, but could you add a docstring to this?  Thanks!

Reply:
In particular, I'd like to know if this is only available in Clarity2, and whether or not other modules need this.  If it's only called from `trait_check_trait_compliance`, then can you assert that the `clarity_version` is Clarity2?  Or even eliminate that argument altogether and indicate in the function name that this is a Clarity2-only function?

The reason for this ask is that I'm trying to identify all the new code that should only affect Clarity2 rules for trait processing, so we can build confidence in the idea that Clarity1 code will be handled the exact same way in `next` as it is in 2.05.